<a href="https://colab.research.google.com/github/shrutimhh/SQL-codes/blob/main/exercise_8_analyzing_sales_data_with_sql.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# 1. Import the packages
import sqlite3 #imports the sqlite package
import random #allows us to generate random variables
from datetime import datetime, timedelta #allows us to analyze date and time values
from google.colab import drive

In [ ]:
# 2. Mount your Google Drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# 3. Connect to the SQLite database in your Google Drive
database_path = "/content/drive/MyDrive/db/sales_data.db"
conn = sqlite3.connect(database_path)
cursor = conn.cursor()

In [ ]:
# 4. Create the sales table
cursor.execute("""
CREATE TABLE sales (
  sales_id INTEGER PRIMARY KEY,
  units_sold INTEGER NOT NULL,
  sales_amount REAL NOT NULL,
  state TEXT NOT NULL,
  dealership TEXT,
  channel TEXT NOT NULL,
  product_id INTEGER NOT NULL,
  sales_date DATE NOT NULL
);
""")

In [ ]:
# 4b. Sample data
states = ['CA', 'TX', 'NY', 'FL', 'IL']
dealerships = ['DealerA', 'DealerB', 'DealerC', 'DealerD', None] #None represents internet sales
channels = ['internet', 'in-person', 'phone']
products = [101, 102, 103, 104, 105]

In [ ]:
# 4c. Insert random data into the table
for _ in range(1000):  # inserting 1000 rows of random data
  units = random.randint(1, 100)
  amount = units * random.uniform(10, 100)
  state = random.choice(states)
  dealership = random.choice(dealerships)
  channel = random.choice(channels)
  product = random.choice(products)
  date = datetime.today() - timedelta(days=random.randint(0, 365 * 3)) #sales dates spanning the last 3 years

  cursor.execute("INSERT INTO sales (units_sold, sales_amount, state, dealership, channel, product_id, sales_date) VALUES (?,?,?,?,?,?,?)",
               (units, amount, state, dealership, channel, product, date.date()))

conn.commit()
conn.close()

In [ ]:
# 5. Load the sql extension
%load_ext sql

In [ ]:
# 6. Connecting to the database using SQL magic for further operations
%sql sqlite:////content/drive/MyDrive/db/sales_data.db

In [ ]:
%%sql
SELECT *
FROM sales

 * sqlite:////content/drive/MyDrive/db/sales_data.db
Done.


[(1, 31, 892.6000297073305, 'NY', 'DealerB', 'internet', 105, '2020-10-07'),
 (2, 79, 2842.7975407225513, 'TX', 'DealerA', 'phone', 101, '2021-07-06'),
 (3, 24, 630.7067058276895, 'IL', 'DealerD', 'internet', 104, '2021-10-15'),
 (4, 62, 707.0002063392164, 'FL', 'DealerB', 'phone', 102, '2023-01-04'),
 (5, 75, 3077.9277129859033, 'NY', 'DealerB', 'in-person', 105, '2023-02-06'),
 (6, 69, 1808.8700557378056, 'NY', 'DealerD', 'in-person', 103, '2022-04-10'),
 (7, 73, 3612.7264279943934, 'IL', 'DealerD', 'phone', 103, '2021-03-10'),
 (8, 93, 7257.219851113585, 'TX', None, 'phone', 101, '2020-12-08'),
 (9, 58, 4115.249487238132, 'TX', 'DealerD', 'internet', 103, '2021-04-19'),
 (10, 66, 5263.560741174115, 'FL', 'DealerD', 'internet', 101, '2023-06-14'),
 (11, 59, 5200.94880405225, 'TX', 'DealerD', 'phone', 102, '2022-09-30'),
 (12, 97, 2749.9375419168614, 'TX', 'DealerA', 'in-person', 101, '2021-01-28'),
 (13, 85, 2329.502711992618, 'CA', 'DealerA', 'phone', 103, '2023-01-11'),
 (14, 71, 6223.790503246842, 'FL', 'DealerB', 'in-person', 104, '2022-01-05'),
 (15, 95, 2422.8547516992344, 'CA', None, 'internet', 102, '2023-07-22'),
 (16, 53, 2363.0297647041853, 'TX', 'DealerD', 'in-person', 101, '2023-02-07'),
 (17, 51, 2211.204748734115, 'CA', 'DealerC', 'in-person', 102, '2021-01-31'),
 (18, 69, 3487.0011031511935, 'TX', 'DealerA', 'internet', 101, '2022-03-13'),
 (19, 99, 9372.536861125975, 'TX', None, 'in-person', 101, '2020-12-15'),
 (20, 52, 2293.2828084491184, 'CA', 'DealerB', 'phone', 101, '2021-06-09'),
 (21, 61, 2738.8347005924097, 'FL', 'DealerD', 'internet', 102, '2021-10-09'),
 (22, 98, 6738.223955824547, 'CA', None, 'internet', 101, '2021-10-01'),
 (23, 90, 6260.668247092346, 'FL', None, 'in-person', 104, '2020-12-27'),
 (24, 29, 1078.9492638580205, 'NY', 'DealerC', 'internet', 102, '2022-01-20'),
 (25, 69, 3252.628578269455, 'TX', 'DealerC', 'internet', 104, '2023-04-22'),
 (26, 6, 268.2798478157872, 'FL', None, 'in-person', 104, '2023-04-04'),
 (27, 92, 7582.366956449473, 'TX', 'DealerC', 'in-person', 101, '2022-01-20'),
 (28, 68, 2096.0781094698477, 'CA', 'DealerA', 'in-person', 103, '2021-02-01'),
 (29, 19, 1269.4015881643593, 'NY', 'DealerC', 'in-person', 103, '2021-03-14'),
 (30, 36, 1887.323677565089, 'TX', 'DealerD', 'in-person', 102, '2021-05-18'),
 (31, 13, 157.35733846770484, 'IL', 'DealerD', 'phone', 103, '2021-12-06'),
 (32, 26, 800.6795248568137, 'TX', 'DealerA', 'in-person', 104, '2022-01-25'),
 (33, 22, 1993.7675580553848, 'TX', 'DealerA', 'internet', 103, '2021-07-05'),
 (34, 93, 5020.4911870388605, 'IL', 'DealerB', 'internet', 104, '2021-11-12'),
 (35, 45, 3412.5964969855854, 'TX', 'DealerB', 'in-person', 105, '2022-10-08'),
 (36, 21, 778.7209340313059, 'NY', 'DealerB', 'phone', 102, '2020-09-30'),
 (37, 30, 469.40362124414617, 'TX', 'DealerB', 'in-person', 102, '2022-05-01'),
 (38, 92, 8121.876285578139, 'CA', None, 'phone', 105, '2021-04-17'),
 (39, 75, 3642.8940438480236, 'IL', 'DealerA', 'internet', 101, '2022-10-30'),
 (40, 60, 4496.5317119210295, 'CA', 'DealerD', 'internet', 102, '2022-01-12'),
 (41, 67, 3428.368351319204, 'TX', 'DealerB', 'phone', 105, '2021-08-29'),
 (42, 58, 1207.555255927429, 'IL', 'DealerC', 'phone', 104, '2023-02-04'),
 (43, 7, 308.8285656082214, 'TX', 'DealerD', 'in-person', 104, '2021-08-21'),
 (44, 65, 4331.460958271848, 'FL', 'DealerD', 'internet', 103, '2022-01-14'),
 (45, 83, 2342.219485867507, 'CA', 'DealerD', 'phone', 103, '2021-08-12'),
 (46, 23, 1978.6343691818647, 'NY', 'DealerD', 'in-person', 103, '2022-04-18'),
 (47, 8, 631.4995492296724, 'IL', None, 'internet', 103, '2022-03-19'),
 (48, 24, 900.8566238837893, 'IL', 'DealerD', 'phone', 104, '2022-07-02'),
 (49, 55, 2870.0513796147693, 'FL', 'DealerB', 'phone', 104, '2022-08-21'),
 (50, 67, 1606.5570355189557, 'TX', 'DealerB', 'internet', 103, '2021-01-28'),
 (51, 96, 9284.614323039572, 'IL', 'DealerB', 'phone', 105, '2021-08-17'),
 (52, 67, 768.2861355052905, 'TX', 'DealerA', 'in-person', 104, '2021-10-14'),
 (5

In [ ]:
%%sql
-- Calculate the total number of unit sales the company has made?
SELECT SUM(units_sold)
FROM sales;

 * sqlite:////content/drive/MyDrive/db/sales_data.db
Done.


[(51800,)]

From the total number of unit sale the company has made, it looks like they had sold quite a lot
-- Interpretation: In the period from 2020 to 2023, the total number of units sold was 51800

In [ ]:
%%sql
-- Calculate the total sales amount in dollars for each state. Your result should be in descending order.
SELECT state, ROUND(SUM(sales_amount),2) as total_sales
FROM sales
GROUP BY state
ORDER BY (total_sales) DESC;

 * sqlite:////content/drive/MyDrive/db/sales_data.db
Done.


[('FL', 666990.7),
 ('IL', 647149.77),
 ('CA', 480799.73),
 ('TX', 445559.76),
 ('NY', 623417.35)]

-- Interpretation: In our data set, Florida had the highest amount of sales. The sales amount of the states are Florida had a total sales of 666,990.7, Illinois had a total of 647,149.77, California had a total of 480,799.73, Texas had a total of 445,559.76, and New York had a total of 623,417.35.

In [ ]:
%%sql
-- Identify the top best dealerships by "State" in terms of the most units sold (ignore internet sales). Your result should be in descending order. Do you notice anything about the state column?
SELECT state, dealership, SUM(units_sold)
FROM sales
WHERE NOT channel = 'internet' AND NOT dealership = 'none'
GROUP BY state, dealership
ORDER BY SUM(units_sold) DESC;

 * sqlite:////content/drive/MyDrive/db/sales_data.db
Done.


[('IL', 'DealerB', 1947),
 ('FL', 'DealerB', 1935),
 ('FL', 'DealerD', 1846),
 ('FL', 'DealerC', 1742),
 ('NY', 'DealerD', 1721),
 ('IL', 'DealerD', 1603),
 ('IL', 'DealerC', 1592),
 ('IL', 'DealerA', 1458),
 ('NY', 'DealerC', 1446),
 ('TX', 'DealerB', 1393),
 ('CA', 'DealerC', 1280),
 ('TX', 'DealerD', 1275),
 ('CA', 'DealerD', 1269),
 ('NY', 'DealerB', 1262),
 ('CA', 'DealerB', 1105),
 ('TX', 'DealerA', 1099),
 ('CA', 'DealerA', 1098),
 ('NY', 'DealerA', 1015),
 ('FL', 'DealerA', 931),
 ('TX', 'DealerC', 503)]

-- Interpretation: We can say that the three best dealerships in terms of units sold are Dealer B in Illinois and Florida, and Dealer D in Florida.  We notice in the state column that there are multiple entries for the same state because the states have multiple dealerships.